### PR TITLE
Compilation fixes on Android

### PR DESCRIPTION
**Pull request by pouleyKetchoupp**
Several compilation fixes on Android platform.

NB: Some changes around Vulkan support are still needed for the codebase to compile and link.

List of fixed errors:
- typeid usage with compilation flag `-fno-rtti`
```
core/rid_owner.h:301:80: error: use of typeid requires -frtti
    print_error("ERROR: " + itos(alloc_count) + " RID allocations of type '" + typeid(T).name() + "' were leaked at exit.");
```

- GLES2 initialization code when OpenGL is disabled
```
drivers/gles2/rasterizer_storage_gles2.h:41:10: fatal error: 'shaders/copy.glsl.gen.h' file not found
    #include "shaders/copy.glsl.gen.h"
```

- Mutex changes (requires c++17 for template argument deduction)
```
servers/visual/visual_server_wrap_mt.h:122:2: error: use of class template 'MutexLock' requires template arguments
    FUNCRID(shader)
servers/server_wrap_mt_common.h:60:4: note: expanded from macro 'FUNCRID'
    MutexLock lock(alloc_mutex);
core/os/mutex.h:60:7: note: template is declared here
    class MutexLock {
```

- Vector changes
```
platform/android/java_godot_lib_jni.cpp:190:50: error: member reference base type 'const int *' is not a structure or union
    env->SetIntArrayRegion(arr, 0, array.size(), r.ptr());      
platform/android/java_godot_lib_jni.cpp:310:4: error: member reference base type 'uint8_t *' (aka 'unsigned char *') is not a structure or union
    w.release();
```

- ObjectDB changes
```
platform/android/java_godot_lib_jni.cpp:1353:39: error: no viable conversion from 'jint' (aka 'int') to 'ObjectID'
    Object *obj = ObjectDB::get_instance(ID);
core/object_id.h:41:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'jint' (aka 'int') to 'const ObjectID &' for 1st argument
    class ObjectID {
core/object_id.h:41:7: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'jint' (aka 'int') to 'ObjectID &&' for 1st argument
core/object.h:790:55: note: passing argument to parameter 'p_instance_id' here
    _ALWAYS_INLINE_ static Object *get_instance(ObjectID p_instance_id) {
```

### CODE REVIEW

**Review by m4gr3d — APPROVED** [review]

File: platform/android/detect.py [hunk 1]
```diff
@@ -205,7 +205,7 @@ def mySpawn(sh, escape, cmd, args, env):
 
     env.Append(CPPFLAGS=["-isystem", env["ANDROID_NDK_ROOT"] + "/sources/cxx-stl/llvm-libc++/include"])
     env.Append(CPPFLAGS=["-isystem", env["ANDROID_NDK_ROOT"] + "/sources/cxx-stl/llvm-libc++abi/include"])
-    env.Append(CXXFLAGS=["-std=gnu++14"])
+    env.Append(CXXFLAGS=["-std=c++17"])
```
Comment (akien-mga) [inline platform/android/detect.py]: We set `gnu++17` in `SConstruct` already, is there any reason to override it and change to `c++17`?

Comment (pouleyKetchoupp) [reply platform/android/detect.py]: No, I just picked `c++17` by mistake. I'm going to remove this line.

File: core/rid_owner.h [hunk 2]
```diff
@@ -298,7 +298,11 @@ class RID_Alloc : public RID_AllocBase {
 			if (description) {
 				print_error("ERROR: " + itos(alloc_count) + " RID allocations of type '" + description + "' were leaked at exit.");
 			} else {
+#if defined(_CPPRTTI)
```
Comment (akien-mga) [inline core/rid_owner.h]: I think we'd usually use `NO_SAFE_CAST` (defined for Android and Web) for this (here `#if !defined(NO_SAFE_CAST)`, or changing the order).

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**